### ISSUE TITLE
TouchID attestation

### ISSUE THREAD

**Issue by Firstyear** [kanidm/webauthn-rs#33]
https://pr-preview.s3.amazonaws.com/alanwaketan/webauthn/pull/1491.html#sctn-apple-anonymous-attestation

This has been documented now, so we should add support for this.

**Comment by ericmarkmartin** [comment]
I was looking into implementing this, and besides not yet having extensions #8 I was wondering how you might go about carrying out step 5 of [the verification procedure given by Apple](https://webkit.org/blog/11312/meet-face-id-and-touch-id-for-the-web/).

> 5. Verify credential public key matches the Subject Public Key of credCert.

Is there a mechanism for comparing public keys? Especially if one of them is a `COSEKey` and the other is an `X509PublicKey`.

**Comment by ericmarkmartin** [comment]
I can throw up a (non-building) draft PR if that would help clarify the question

**Comment by Firstyear** [comment]
I think I can see the question, but forgive if answer is unclear. Reading this, the Subject Public Key of credCert doesn't seem to be defined in the signing procedure, so it does appear to be ambiguous what should be carried out in step 5. Having access to a credCert X509 that we can inspect would probably be helpful here to understand what is actually represented in the data. 

If @alanwaketan has some time, would you mind clarifying in step 4 of the Apple Anonymous Attestation how the Subject Public Key is constructed for verification in step 5 of the verification procedure? 

Thanks,

**Comment by ericmarkmartin** [comment]
I was trying to look to the change adding support for Apple's anonymous CA attestation in Duo Labs' py_webauthn [here](https://github.com/duo-labs/py_webauthn/pull/85/files) for inspiration but it wasn't clear to me how to replicate that work with the rust openssl library.

**Comment by aseigler** [comment]
> I think I can see the question, but forgive if answer is unclear. Reading this, the Subject Public Key of credCert doesn't seem to be defined in the signing procedure, so it does appear to be ambiguous what should be carried out in step 5. Having access to a credCert X509 that we can inspect would probably be helpful here to understand what is actually represented in the data.
> 
> If @alanwaketan has some time, would you mind clarifying in step 4 of the Apple Anonymous Attestation how the Subject Public Key is constructed for verification in step 5 of the verification procedure?
> 
> Thanks,

From https://tools.ietf.org/html/rfc5480#section-2.2, "The subjectPublicKey from [SubjectPublicKeyInfo](https://tools.ietf.org/html/rfc5280#section-4.1.2.7) is the ECC public key."

The way I implemented this part is a bit roundabout but I already had the plumbing in place so it was a cinch:

1. Extract [COSE algorithm](https://www.iana.org/assignments/cose/cose.xhtml#algorithms) from credential public key (obtained from AttestedCredentialData from within authenticatorData).  Observed to be ECDSA w/ SHA-256, which is -7.
2. Extract public key type from credCert.  Observed to be "1.2.840.10045.2.1" aka [id-ecPublicKey](https://tools.ietf.org/html/rfc5480#section-2.1) which maps to [COSE key type EC2 value](https://www.iana.org/assignments/cose/cose.xhtml#key-type) 2.
3. Extract [ECParameters](https://tools.ietf.org/html/rfc5480#section-2.1) from credCert.  
4. Extract curve field from ECParameters, observed to be "1.2.840.10045.3.1.7" aka secp256r1 which maps to [COSE elliptic curve value](https://www.iana.org/assignments/cose/cose.xhtml#elliptic-curves) P-256, which is 1.
5. Extract ECPoint field from ECParameters and extract X and Y coordinates from the ECPoint, which map to [COSE key type parameter values](https://www.iana.org/assignments/cose/cose.xhtml#key-type-parameters) -2 and -3 respectively.
6. With the COSE key type, algorithm, curve, X, and Y values, build a new [credential public key](https://www.w3.org/TR/webauthn/#sctn-encoded-credPubKey-examples) and perform CBOR encoding on that.
7. Compare the credential public key from AttestedCredentialData from the authenticatorData with the credential public key built.  They should match.

My initial implementation [here](https://github.com/abergs/fido2-net-lib/blob/0d6410c4ad3b1103eb5a62cc85acc8b9e2a9c4af/Src/Fido2/AttestationFormat/Apple.cs#L81) and building the credential public key [here](https://github.com/abergs/fido2-net-lib/blob/0d6410c4ad3b1103eb5a62cc85acc8b9e2a9c4af/Src/Fido2/Objects/CredentialPublicKey.cs#L29) .

**Comment by ericmarkmartin** [comment]
Does OpenSSL rust provide a means of getting at the subject public key? It seems the [x509-parser crate does](https://docs.rs/x509-parser/0.9.0/x509_parser/x509/struct.SubjectPublicKeyInfo.html#structfield.subject_public_key)

**Comment by Firstyear** [comment]
> Does OpenSSL rust provide a means of getting at the ECParameters in the credCert? I couldn't seem to find anything

Yes it does - I'd rather not pull in any other crypto libs that OpenSSL at the moment (I want to swap to ring, but it has no way to reconstruct keys from parameters atm). 

For now, my advice would be to capture a touch id registration to create a test case, and then implement everything except step 5. We'll review the first part, then we can focus on the step 5 verification since it will probably need a bit more care to get right. How's that sound?

PS: The due labs impl often is a good reference to look at https://github.com/duo-labs/webauthn 

Thanks for chiming in @aseigler :)

**Comment by ericmarkmartin** [comment]
> Yes it does - I'd rather not pull in any other crypto libs that OpenSSL at the moment (I want to swap to ring, but it has no way to reconstruct keys from parameters atm).

Interesting—out of curiosity do you have a link to the docs?
 
> For now, my advice would be to capture a touch id registration to create a test case, and then implement everything except step 5. We'll review the first part, then we can focus on the step 5 verification since it will probably need a bit more care to get right. How's that sound?

I think we're also unable to do step 4 as of yet because we are missing extensions.

> PS: The due labs impl often is a good reference to look at https://github.com/duo-labs/webauthn

I was looking at their python implementation because my complete inexperience with go. I think I linked the relevant bit of code from the python earlier in this thread.

**Comment by Firstyear** [comment]
> > Yes it does - I'd rather not pull in any other crypto libs that OpenSSL at the moment (I want to swap to ring, but it has no way to reconstruct keys from parameters atm).
> 
> Interesting—out of curiosity do you have a link to the docs?

From https://docs.rs/openssl/0.10.30/openssl/pkey/struct.PKeyRef.html#method.ec_key you can get https://docs.rs/openssl/0.10.30/openssl/ec/struct.EcKey.html#method.public_key and https://docs.rs/openssl/0.10.30/openssl/ec/struct.EcGroupRef.html which you can the construct into a COSEKey which can be compared directly. 

> 
> > For now, my advice would be to capture a touch id registration to create a test case, and then implement everything except step 5. We'll review the first part, then we can focus on the step 5 verification since it will probably need a bit more care to get right. How's that sound?
> 
> I think we're also unable to do step 4 as of yet because we are missing extensions.

That's fine. For now lets get the fundamentals in place then we can look at step 4 and 5. 

> 
> > PS: The due labs impl often is a good reference to look at https://github.com/duo-labs/webauthn
> 
> I was looking at their python implementation because my complete inexperience with go. I think I linked the relevant bit of code from the python earlier in this thread.

I have no experience in go at all either, and can barely read it, but in terms of just having the steps listed there it can be useful.

**Comment by ericmarkmartin** [comment]
See #41
